File: src/main/java/com/pauljoda/modularsystems/power/tiles/TileBankRF.java
package com.pauljoda.modularsystems.power.tiles;

import cofh.api.energy.EnergyStorage;
import cofh.api.energy.IEnergyHandler;
import com.pauljoda.modularsystems.core.registries.ConfigRegistry;
import com.pauljoda.modularsystems.power.gui.GuiRFBank;
import com.teambr.bookshelf.common.tiles.IOpensGui;
import com.teambr.bookshelf.helpers.GuiHelper;
import com.teambr.bookshelf.inventory.ContainerGeneric;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.world.World;
import net.minecraftforge.common.util.ForgeDirection;

import java.util.List;

public class TileBankRF extends TileBankBase implements IOpensGui, IEnergyHandler {

    private EnergyStorage energy;

    public TileBankRF() {
        energy = new EnergyStorage(10000);
    }

    @Override
    public double getPowerLevelScaled(int scale) {
        return energy.getEnergyStored() * scale / energy.getMaxEnergyStored();
    }

    /*******************************************************************************************************************
     ***************************************** Fuel Provider Functions *************************************************
     *******************************************************************************************************************/

    @Override
    public boolean canProvide() {
        return energy.getEnergyStored() > 0;
    }

    @Override
    public double fuelProvided() {
        int actual = energy.extractEnergy((int)Math.round(ConfigRegistry.rfPower * 200), true);
        return (actual / (ConfigRegistry.rfPower * 200)) * 200;
    }

    @Override
    public double consume() {
        int actual = energy.extractEnergy((int)Math.round(ConfigRegistry.rfPower * 200), false);
        worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
        return (actual / (ConfigRegistry.rfPower * 200)) * 200;
    }

    @Override
    public FuelProviderType type() {
        return FuelProviderType.RF;
    }

    /*******************************************************************************************************************
     ****************************************** Energy Methods *********************************************************
     *******************************************************************************************************************/

    @Override
    public int receiveEnergy(ForgeDirection from, int maxReceive, boolean simulate) {
        if (getCore() != null) {
            int actual = energy.receiveEnergy(maxReceive, simulate);
            worldObj.markBlockForUpdate(xCoord, yCoord, zCoord);
            return actual;
        }
        return 0;
    }

    @Override
    public int extractEnergy(ForgeDirection from, int maxExtract, boolean simulate) {
        return 0;
    }

    @Override
    public int getEnergyStored(ForgeDirection forgeDirection) {
        return energy.getEnergyStored();
    }

    @Override
    public int getMaxEnergyStored(ForgeDirection forgeDirection) {
        return energy.getMaxEnergyStored();
    }

    @Override
    public boolean canConnectEnergy(ForgeDirection from) {
        return true;
    }

    /*******************************************************************************************************************
     ******************************************** Tile Methods *********************************************************
     *******************************************************************************************************************/

    @Override
    public void readFromNBT (NBTTagCompound tags) {
        super.readFromNBT(tags);
        energy.readFromNBT(tags);
    }

    @Override
    public void writeToNBT (NBTTagCompound tags) {
        super.writeToNBT(tags);
        energy.writeToNBT(tags);
    }

    /*******************************************************************************************************************
     ****************************************** IOpensGui Methods ******************************************************
     *******************************************************************************************************************/

    @Override
    public Object getServerGuiElement(int i, EntityPlayer entityPlayer, World world, int i1, int i2, int i3) {
        return new ContainerGeneric();
    }

    @Override
    public Object getClientGuiElement(int i, EntityPlayer entityPlayer, World world, int i1, int i2, int i3) {
        return new GuiRFBank(this);
    }

    /*******************************************************************************************************************
     ************************************************ Waila ************************************************************
     *******************************************************************************************************************/

    @Override
    public void returnWailaHead(List<String> list) {
        list.add(GuiHelper.GuiColor.YELLOW + "Available Power: " + GuiHelper.GuiColor.WHITE + energy.getEnergyStored() + "/" + energy.getMaxEnergyStored());
    }
}

File: src/main/scala/com/teambr/modularsystems/power/tiles/TileBankRF.scala
package com.teambr.modularsystems.power.tiles

import com.teambr.bookshelf.common.tiles.traits.EnergyHandler
import com.teambr.modularsystems.core.providers.FuelProvider.FuelProviderType
import com.teambr.modularsystems.core.registries.ConfigRegistry
import net.minecraft.nbt.NBTTagCompound

/**
 * This file was created for Modular-Systems
 *
 * Modular-Systems is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Dyonovan
 * @since August 10, 2015
 */
class TileBankRF extends TileBankBase with EnergyHandler {

    /**
     * Used to scale the current power level
      *
      * @param scale The scale to move to
     * @return A number from 0 - { @param scale} for current level
     */
    override def getPowerLevelScaled(scale: Int): Double = energyStorage.getEnergyStored * scale / energyStorage.getMaxEnergyStored

    /*******************************************************************************************************************
      ******************************************** RF Energy Functions *************************************************
      ******************************************************************************************************************/

    override def defaultEnergyStorageSize: Int = 80000

    override def isReceiver: Boolean = true

    override def isProvider: Boolean = false

    /*******************************************************************************************************************
      **************************************** Fuel Provider Functions *************************************************
      ******************************************************************************************************************/

    override def consume(): Double = {
        val actual = energyStorage.extractEnergy(Math.round(ConfigRegistry.rfPower * 200).toInt, false)
        worldObj.notifyBlockUpdate(pos, worldObj.getBlockState(pos), worldObj.getBlockState(pos), 6)
        (actual / (ConfigRegistry.rfPower * 200)) * 200
    }

    override def `type`(): FuelProviderType = FuelProviderType.RF

    override def fuelProvided(): Double = {
        val actual = energyStorage.extractEnergy(Math.round(ConfigRegistry.rfPower * 200).toInt, false)
        (actual / (ConfigRegistry.rfPower * 200)) * 200
    }

    override def canProvide: Boolean = energyStorage.getEnergyStored > 0

    /*******************************************************************************************************************
      ******************************************** Tile Methods ********************************************************
      ******************************************************************************************************************/
    override def readFromNBT (tags: NBTTagCompound) {
        super[TileBankBase].readFromNBT(tags)
        super[EnergyHandler].readFromNBT(tags)
    }


    override def writeToNBT(tags: NBTTagCompound) : NBTTagCompound = {
        super[TileBankBase].writeToNBT(tags)
        super[EnergyHandler].writeToNBT(tags)
        tags
    }
}
